### PR TITLE
Improve command line flags and usage message

### LINKED ISSUE #195: Improve command line flags and usage message
Students are generally clueless on how to use OK other than copy-pasting in commands. We should direct them to use `python3 ok --help` in our course introduction, and improve that message so it's actually useful.

Here's the current output:
```
usage: ok [-h] [-q QUESTION] [-u] [-i] [-v] [--all] [--submit] [--backup]
          [--revise] [--restore] [--lock] [--score] [--score-out SCORE_OUT]
          [--config CONFIG] [--timeout TIMEOUT] [--no-hints] [--hint]
          [--style] [--export] [--latest] [--version] [--tests] [--debug]
          [--local] [--server SERVER] [--authenticate] [--get-token]
          [--insecure] [--no-update] [--update]

This file is responsible for coordinating all of OK's protocols.

optional arguments:
  -h, --help            show this help message and exit
  -q QUESTION, --question QUESTION
                        focus on specific questions
  -u, --unlock          unlock tests interactively
  -i, --interactive     toggle interactive mode
  -v, --verbose         print more output
  --all                 run tests for all questions in config file
  --submit              Submit assignment
  --backup              Backup assignment reliably
  --revise              Submit composition revision
  --restore             Restore assignment from an earlier backup
  --lock                partial path to directory to lock
  --score               Scores the assignment
  --score-out SCORE_OUT
                        file to write scores to
  --config CONFIG       Specify a configuration file
  --timeout TIMEOUT     set the timeout duration for running tests
  --no-hints            Do not run Hinting Protocol
  --hint                Run hinting protocol and compute hint (if available)
  --style               Run AutoStyle Feedback System
  --export              Downloads all submissions for the current assignment
  --latest              When used with --export, downloads latest submissions
                        instead of final submissions
  --version             Prints the version number and quits
  --tests               display a list of all available tests
  --debug               show debug statements
  --local               disable any network activity
  --server SERVER       server address
  --authenticate        authenticate, ignoring previous authentication
  --get-token           gets ok access token
  --insecure            uses http instead of https
  --no-update           turns off software updating
  --update              checks and performs software update then exits
```

I see a few ways this could be improved:
* The usage message is way too long. It should something like `python3 ok [--help] [OPTIONS]` (or e.g. `py ok` if that's how they invoked it).
* "This file is responsible for coordinating all of OK's protocols" should be replaced with a better description what the program actually does.
* The flags should be grouped so that ones that students are likely to use (`--help`, `--submit`, `-q`, `-u`, `--local`, `--interactive`) are at the top. Debugging or grading flags (`--insecure`, `--score`, etc.) should be last.

### PR DESCRIPTION
**Pull request by knrafto**
Resolves #195.

Here's the new help message:
```
usage: python3 ok [--help] [options]

ok is an autograder that you can use to run tests, back up your work, and
submit assignments.

You can run all tests with

    python3 ok

There are several "options" you can give ok to modify its behavior. These
options generally have both a short form (preceded by a single dash, like -q)
or a long form (preceded by two dashes, like --question). This is similar to how
many other command line applications accept options. These options can be mixed
and matched in any order. The options are listed in full below, but we'll
describe some of the more common ones here.

To test a specific question, use the -q (or --question) option with the name of
the question:

    python3 ok -q foo
    python3 ok -q 12

By default, only tests that fail will appear. If you want to see the results
from all tests, you can use the -v (or --verbose) option:

    python3 ok -q foo -v

To start an interactive interpreter after a failed test for debugging, use the
-i (or --interactive) option:

    python3 ok -q foo -i

By default, after each test run ok will attempt to back up your work to the
server. To run the tests without any network access, use the --local option:

    python3 ok -q foo --local

To submit the assignment after you're done, use the --submit option:

    python3 ok --submit

Finally, to log out and log in under a different email, use --authenticate:

    python3 ok --authenticate

Visit https://okpy.org to view your backups and submissions.

optional arguments:
  -h, --help            show this help message and exit

running tests:
  -q QUESTION, --question QUESTION
                        run tests for a specific question
  -u, --unlock          unlock tests interactively
  -i, --interactive     start the Python interpreter after a failed test
  -v, --verbose         show all tests, not just passing tests
  --all                 run tests for all questions in config file
  --submit              submit the assignment
  --backup              attempt to reliably backup your work
  --revise              submit composition revision
  --restore             restore assignment from an earlier backup
  --timeout TIMEOUT     set the timeout duration (in seconds) for running
                        tests

experiment options:
  --no-hints            do not give hints
  --hint                give a hint (if available)
  --style               run AutoStyle feedback system

debugging options:
  --version             print the version number and exit
  --tests               display a list of all available tests
  --debug               show debugging output

grading options:
  --lock                lock the tests in a directory
  --score               score the assignment
  --score-out SCORE_OUT
                        write scores to a file
  --config CONFIG       use a specific configuration file
  --export              download all submissions for the current assignment
  --latest              when used with --export, download latest submissions
                        instead of final submissions

server options:
  --local               disable any network activity
  --server SERVER       set the server address
  --authenticate        authenticate, ignoring previous authentication
  --get-token           get ok access token
  --insecure            use http instead of https
  --no-update           do not check for ok updates
  --update              update ok and exit
```

I'm worried the initial message might be too long, but I tried to be friendly to those who have never used a command line before.

Nitpicks welcome!

### CODE REVIEW

File: client/cli/ok.py [hunk 1]
```diff
@@ -27,82 +67,85 @@
...
         description=__doc__,
+        usage='%(prog)s [--help] [options]',
         formatter_class=argparse.RawDescriptionHelpFormatter)
-    # Protocol parameters
-    parser.add_argument('-q', '--question', type=str, action='append',
-                        help="focus on specific questions")
-    parser.add_argument('-u', '--unlock', action='store_true',
+
+    testing = parser.add_argument_group('running tests')
+    testing.add_argument('-q', '--question', type=str, action='append',
```
Comment (Sumukh) [inline client/cli/ok.py]: typo: specifi

**Comment by Sumukh** [comment]
A common request is how to login under a different email. We might want add that to the blurb (it's currently at the bottom as one of many flags) 

I think most of the info from http://cs61a.org/articles/using-ok.html is here. 

Another related idea: 
Have a sample assignment with an unlocking test suite that serves as an ok walkthrough. Instructors could use that as their intro to using Ok assignment (for 61A: that would be in Lab 0)

**Comment by knrafto** [comment]
I've fixed a couple typos and updated the PR description. I did take inspiration from the "Using OK" article, so hopefully we can direct students to this help message instead.

File: client/cli/ok.py [hunk 2]
```diff
@@ -27,82 +71,85 @@
...
+    grading = parser.add_argument_group('grading options')
+    grading.add_argument('--lock', action='store_true',
+                        help="lock the tests in a directory")
+    grading.add_argument('--score', action='store_true',
+                        help="score the assignment")
+    grading.add_argument('--score-out', type=argparse.FileType('w'),
+                        default=sys.stdout, help="write scores to a file")
+    grading.add_argument('--config', type=str,
+                        help="use a specific configuration file")
+    grading.add_argument('--export', action='store_true',
```
Comment (Sumukh) [inline client/cli/ok.py]: `--export` is broken for v3. Can we remove it for now? Hopefully, we'll get it back at some point.

Comment (knrafto) [inline client/cli/ok.py]: Sure. Should this functionality even be in ok-client?

**Comment by knrafto** [comment]
Similar to `--export`, should we remove `--lock` as well? Instead, we should use `ok-lock`, like the 61A repo does.

**Comment by Sumukh** [comment]
I think students actually use `--lock` (at least I've told a few students when they want to redo a test suite) - so I'm inclined to keep that one.

**Comment by knrafto** [comment]
^ Tested manually too by adding a bogus protocol to one of the demos.

File: client/protocols/backup.py [hunk 3]
```diff
@@ -24,7 +24,7 @@ class BackupProtocol(models.Protocol):
     REVISION_ENDPOINT = '{prefix}://{server}/api/v3/revision/'
 
     def run(self, messages):
-        if self.args.local or self.args.export or self.args.restore:
+        if self.args.local or self.args.restore:
```
Comment (Sumukh) [inline client/protocols/backup.py]: Not in the scope of this PR - but I wonder if there's a better to way to specify when to not run a protocol. We repeat this logic in all protocols.

Maybe all protocols should have a method called `can_run(self)` that returns true or false before we attempt to run them.

**Comment by Sumukh** [comment]
LGTM.

**Pull request merged**